### PR TITLE
Add methods for converting to/from a raw `acl_t` pointer

### PR DESCRIPTION
**Pull request by aidanhs**

### CODE REVIEW

**Comment by aidanhs** [comment]
Moving discussion from https://github.com/intgr/posix-acl/pull/18#discussion_r384107617

> It seems this method should be marked unsafe, what do you think?

I see this as being comparable to [as_raw_fd on File](https://doc.rust-lang.org/stable/std/os/unix/io/trait.AsRawFd.html). But afaik you can't cause UB with an invalid/closed file descriptor, so perhaps it's more similar to [`into_raw` and `from_raw` for Rc](https://doc.rust-lang.org/stable/std/rc/struct.Rc.html#method.into_raw) (note that `into_raw` is a safe function).

I'll have a look at mimicking the Rc API and updating this PR!

> Just wondering, what use case do you have for this? Maybe it makes sense to implement your needed feature directly in posix-acl, in addition to this.

I'm trying to use some nonstandard extensions to POSIX ACLs (`acl_to_any_text` and `acl_extended_file_nofollow`).

To be honest, I think there are diminishing returns with trying to figure out platform support for these different extensions. Taking the Python library as an example, after reviewing the docs for [investigating compatibility for a new platform](https://pylibacl.k1024.org/implementation.html#currently-supported-platforms), combined with the need to then [carve out subsets of the ACL interface](with a compile-time platform check)  and the number of `ifdef`s in the [source](https://github.com/iustin/pylibacl/blob/master/acl.c)...it'd be possible for sure, but it feels like a fair amount of work that I'm not personally enthusiastic to tackle at present :sweat_smile:

**Comment by aidanhs** [comment]
After playing around a bit with `into_raw` and `from_raw`, I kinda prefer the current API i.e. `as_raw` - it makes it easier to make my own extension traits like so:

```rust
        fn check_return(ret: i32, func: &str) {
            if ret != 0 { 
                panic!("Error in {}: {}", func, io::Error::last_os_error());
            }
        }
        struct AutoPtr<T>(*mut T); 
        impl<T> Drop for AutoPtr<T> {
            fn drop(&mut self) {
                let ret = unsafe { acl_sys::acl_free(self.0 as *mut libc::c_void) };
                check_return(ret, "acl_free");
            }
        }
        trait PosixACLToAnyText {
            fn to_any_text(&self, prefix: &[u8], separator: u8, options: libc::c_int) -> Result<Vec<u8>, Void>;
        }
        impl PosixACLToAnyText for PosixACL {
            fn to_any_text(&self, prefix: &[u8], separator: u8, options: libc::c_int) -> Result<Vec<u8>, Void> {
                let c_prefix = CString::new(prefix).unwrap();
                let ret = unsafe { acl_to_any_text(self.as_raw(), c_prefix.as_ptr(), separator as i8, options) };
                if ret.is_null() {
                    panic!(
                        "Error formatting ACL {:?} {} {}: {}",
                        prefix,
                        separator,
                        options,
                        io::Error::last_os_error()
                    );
                }
                let ret = AutoPtr(ret);
                let c_ret = unsafe { CStr::from_ptr(ret.0) };
                Ok(c_ret.to_bytes().to_owned())
            }
        }
```
(ignore `AutoPtr` and panicking instead of error handling: I copied `AutoPtr` out because it was convenient but it's very optional, and I'm just not focused on error handling right now)

Note that the extension trait takes an `&self`, like all of the built in functions do - if I only had `{into,from}_raw` I'd have to do something less seamless.

I could add all three functions (`{as,into,from}_raw`) if you'd prefer? But that may be unnecessary API surface until someone actually needs it.

**Comment by intgr** [comment]
One downside with the `as_raw()` API is that this cements PosixACL to be stuck with using the `acl_t` representation internally, which is inefficient (e.g. addition currently has n^2 complexity because it needs to check all previous entries).

However, I figure ACLs are typically so small that this doesn't matter.

**Comment by intgr** [comment]
> I could add all three functions ({as,into,from}_raw) if you'd prefer? But that may be unnecessary API surface until someone actually needs it.

Yeah, let's not implement those until there's a use case. :)

File: src/lib.rs [hunk 1]
```diff
@@ -337,6 +337,15 @@ impl PosixACL {
         }
         Ok(())
     }
+
+    /// Returns a raw `acl_t` type that can be used directly in FFI calls to the acl library.
+    ///
+    /// The caller *must not*:
+    /// - `acl_free` or otherwise invalidate the returned `acl_t`
+    /// - use this `acl_t` after the PosixACL has been dropped
+    pub fn as_raw(&self) -> acl_sys::acl_t {
```
Comment (intgr) [inline src/lib.rs]: Hmm, can you annotate the returned `acl_t` with the lifetime of `self`, so the compiler can verify the condition "_use this `acl_t` after the PosixACL has been dropped_"

If yes, I don't see any issue with omitting `unsfe`.

**Comment by intgr** [comment]
> To be honest, I think there are diminishing returns with trying to figure out platform support for these different extensions.

After taking a look at `acl.h` from a few different OS, I agree :)

File: src/lib.rs (hunk 1 above)
Comment (aidanhs) [reply src/lib.rs]: I was thinking about this and I don't see any way to do it - ultimately, you're returning a pointer (as that's all an acl_t is).

I've ended up actually convincing myself that `fn into_raw` and `unsafe fn from_raw` are the right way to go - it's a well understood pattern in Rust (present in the stdlib) and hints at the pointer ownership issues you need to keep in mind.

(it also means that you're not locked into an internal acl_t representation forever - you just need to be able to create from from your internal representation)

**Comment by aidanhs** [comment]
I've changed this PR to use `into_raw`/`from_raw` instead, like Rc. The doc comments are written to mimic the Rc methods (as are the unsafe annotations).

File: src/lib.rs (hunk 1 above)
Comment (intgr) [reply src/lib.rs]: Ah my bad, only `&` refs can have lifetimes, not `*` raw pointers.

File: src/lib.rs [hunk 2]
```diff
@@ -437,6 +437,28 @@ impl PosixACL {
...
+
+    /// Consumes the `PosixACL`, returning the wrapped `acl_t`.
+    /// This can be used directly in FFI calls to the acl library.
+    ///
+    /// To avoid a memory leak, the pointer must either:
+    ///
+    /// - Be converted back to a PosixACL using `PosixACL::from_raw`
+    /// - Have `acl_free` called on it
+    pub fn into_raw(&self) -> acl_sys::acl_t {
+        self.acl
```
Comment (intgr) [inline src/lib.rs]: I don't understand how the "Consumes the `PosixACL`" part works. ISTM that Rust will still call `drop` on the original `PosixACL`, which triggers `acl_free()`, meaning that if you pass the same pointer to `PosixACL::from_raw()` it causes double-free.

Also I believe the `into_raw` API should be marked unsafe still, if it's possible to cause a memory leak. This wasn't the case with the previous proposed `as_raw` API.

I wrote this test case which demonstrates the behaviors:
```rust
#[test]
fn acl_into_from_raw() {
    let acl1 = full_fixture();
    let ptr = acl1.into_raw();

    // Should I be able to call `.entries()` after converting acl1 into raw ptr?
    // The docstring suggest no, but this works.
    assert_eq!(acl1.entries().len(), 8);

    // Causes error:
    // thread 'acl_into_from_raw' panicked at 'Error in acl_free: Invalid argument (os error 22)', src/lib.rs:209:9
    let _acl2 = unsafe { PosixACL::from_raw(ptr) };
}
```

File: src/lib.rs [hunk 3]
```diff
@@ -437,6 +437,28 @@ impl PosixACL {
...
     }
+
+    /// Consumes the `PosixACL`, returning the wrapped `acl_t`.
+    /// This can be used directly in FFI calls to the acl library.
+    ///
+    /// To avoid a memory leak, the pointer must either:
+    ///
+    /// - Be converted back to a PosixACL using `PosixACL::from_raw`
+    /// - Have `acl_free` called on it
+    pub fn into_raw(&self) -> acl_sys::acl_t {
```
Comment (intgr) [inline src/lib.rs]: Everywhere else in this file these types are without the `acl_sys::` prefix.

Comment (intgr) [reply src/lib.rs]: Also note the Clippy convention warnings here: https://github.com/intgr/posix-acl/pull/21/checks?check_run_id=501508319

**Review by intgr — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (aidanhs) [reply src/lib.rs]: Yikes! Thanks for catch, I was in a hurry and pushed the wrong version of this - let me re-push  the fixed one in a bit.

On leaking being unsafe - leaking is typically considered safe in rust programs. For example:
 - the safe stdlib function specifically designed to leak memory (https://doc.rust-lang.org/std/mem/fn.forget.html)
 - `Rc::into_raw` also in the stdlib is not unsafe

The nomicon has a more extended discussion on the topic (https://doc.rust-lang.org/nomicon/leaking.html), and the book references it as well (https://doc.rust-lang.org/book/ch15-06-reference-cycles.html#reference-cycles-can-leak-memory)

> Preventing memory leaks entirely is not one of Rust’s guarantees in the same way that disallowing data races at compile time is, meaning memory leaks are memory safe in Rust.

However - as the maintainer of this crate ultimately the decision is up to you! I'll update the PR with to make `into_raw` be unsafe based on your previous feedback.

Comment (intgr) [reply src/lib.rs]: > On leaking being unsafe - leaking is typically considered safe in rust programs.

Ahh thanks for that background. I didn't understand the implications of what you mentioned about `Rc::into_raw` until now. I agree it makes sense to be consistent with Rust conventions. 

Adding a code comment to one of these resources might be helpful. :)

Comment (intgr) [reply src/lib.rs]: Just to be clear: I agree now that `into_raw` shouldn't be `unsafe`.

**Comment by intgr** [comment]
Thanks, looks good to me now, and I learned a few new things about Rust. :)

**Review by intgr — APPROVED** [review]

**Comment by aidanhs** [comment]
Thanks for iterating with me on this!

**Comment by intgr** [comment]
@aidanhs Released this as part of version 0.5.0.

**Pull request merged**